src/server/pkg/hashtree/db.go

```diff
@@ -0,0 +1,844 @@
+package hashtree
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	pathlib "path"
+	"regexp"
+	"strings"
+
+	"golang.org/x/sync/errgroup"
+
+	bolt "github.com/coreos/bbolt"
+	globlib "github.com/gobwas/glob"
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+)
+
+const (
+	FsBucket      = "fs"
+	ChangedBucket = "changed"
+	perm          = 0666
+)
+
+var (
+	changedVal     = []byte{1}
+	emptyStringKey = []byte{0}
+)
+
+func fs(tx *bolt.Tx) *bolt.Bucket {
+	return tx.Bucket(b(FsBucket))
+}
+
+func changed(tx *bolt.Tx) *bolt.Bucket {
+	return tx.Bucket(b(ChangedBucket))
+}
+
+type dbHashTree struct {
+	*bolt.DB
+}
+
+func s(b []byte) string {
+	if bytes.Equal(b, emptyStringKey) {
+		return ""
+	}
+	return string(b)
+}
+
+func b(s string) []byte {
+	if s == "" {
+		return emptyStringKey
+	}
+	return []byte(s)
+}
+
+func dbFile() string {
+	return fmt.Sprintf("/tmp/db/%s", uuid.NewWithoutDashes())
+}
+
+func NewDBHashTree() (HashTree, error) {
```

Comment:
Why are we explicit about this being stored in a DB? Isn't that just an implementation detail that we should abstract away in the interface?